firestore-tos-extension/functions/src/index.ts

```diff
@@ -0,0 +1,245 @@
+
+    if (data.status === AcknowledgementStatus.ACCEPTED) {
+      //@ts-ignore
+      acknowledgementDates.acknowledgedDate =
+        admin.firestore.FieldValue.serverTimestamp();
+    }
+
+    if (data.status === AcknowledgementStatus.DECLINED) {
+      //@ts-ignore
+      acknowledgementDates.unacknowledgedDate =
+        admin.firestore.FieldValue.serverTimestamp();
+    }
+
+    /** Set new acknowledgment */
+    const ack = {
+      ...noticeDoc.data(),
+      ...data,
+      ...acknowledgementDates,
+    };
+
+    /** Add Acknowledgment */
+    await db
+      .collection(config.collectionPath)
+      .doc("acknowledgements")
+      .collection(context.auth.uid)
+      .doc(data.noticeId)
+      .withConverter(acknowledgementConverter)
+      .set(
+        { ...ack },
+        {
+          merge: true,
+        }
+      );
+
+    /** Find current acknowledgements */
+    const acknowledgements = await db
+      .collection(config.collectionPath)
+      .doc("acknowledgements")
+      .collection(context.auth.uid)
+      .withConverter(acknowledgementConverter)
+      .get()
+      .then(($) => $.docs.map((doc) => doc.data()));
+
+    /** Set claims on user and return */
+    const claims = {};
+    claims[process.env.EXT_INSTANCE_ID] = acknowledgements;
+
+    const { customClaims: existingClaims } = await auth.getUser(
+      context.auth.uid
+    );
+
+    const updatedClaims = await auth.setCustomUserClaims(context.auth.uid, {
+      ...existingClaims,
+      ...claims,
+    });
+
+    /** send event if configured */
+    await eventChannel?.publish({
+      type: "firebase.google.v1.acknowledgement-accepted",
+      data: JSON.stringify(updatedClaims),
+    });
+  }
+);
+
+export const createNotice = functions.handler.https.onCall(
+  async (data, context) => {
+    // Checking that the user is authenticated.
+    if (!context.auth) {
+      throw new functions.https.HttpsError(
+        "unauthenticated",
+        "No valid authentication token provided."
+      );
+    }
+
+    /** check if notice data has been provided  */
+    if (!data) {
+      throw new functions.https.HttpsError(
+        "invalid-argument",
+        "No valid data provided."
+      );
+    }
+
+    /** check if notice data has been provided  */
+    if (!data.noticeType) {
+      throw new functions.https.HttpsError(
+        "invalid-argument",
+        "Invalid notice type"
+      );
+    }
+
+    /** Set claims on user and return */
+    return db
+      .collection(config.collectionPath)
+      .doc("agreements")
+      .collection("notices")
+      .doc(data.noticeId)
+      .withConverter(noticeConverter)
+      .set(
+        {
+          ...data,
+          creationDate: admin.firestore.FieldValue.serverTimestamp(),
+        },
+        { merge: true }
+      );
+  }
+);
+
+export const getNotices = functions.handler.https.onCall(
+  async (data, context) => {
+    // Checking that the user is authenticated.
+    if (!context.auth) {
```

Comment:
We probably don't want to require auth here. Imagine a ToS form before the user has been created. However we probably to want to hide results if there is an allow_list and the user doesn't match the allow_list.

Reply:
Would removing auth expose the function for potential abuse. Could users use anonymous authentication as a "user"?

Reply:
Good point, I think you're right. I guess we should probably keep it then.